Iter13

SinnerUfa opened the pull request.

SinnerUfa committed:
Итерация 1 -  сервер + Итерации 4, 5 (флаги и окружение) + вывод страницы со всеми метриками

SinnerUfa committed:
Начало работы с автотестом

SinnerUfa committed:
Исправление ошибок найденных go vet

SinnerUfa committed:
Исправление вновь найденных ошибок go vet

SinnerUfa committed:
Исправление ошибок найденных go vet

SinnerUfa committed:
Проверка влияния sync

SinnerUfa committed:
Итерация 2, сделать проверку автотестами

SinnerUfa committed:
Итер3 добавлено закомментирование формаирование страницы на адресе сервера (необходимо посмотреть тесты)

SinnerUfa committed:
Итер 3 странность с обработкой корневого адреса (проверка завершилась на итер 1) необходим возврат и просмотр тестов

SinnerUfa committed:
Итер 3 обработка float

SinnerUfa committed:
Итер3 проблема float

SinnerUfa committed:
Итер3 проблема float

SinnerUfa committed:
Итер3 проблема сторонних пакетов

SinnerUfa committed:
Итер3 проблема сторонних пакетов

SinnerUfa committed:
Итер3 проблема сторонних пакетов

SinnerUfa committed:
Итер3 проблема сторонних пакетов

SinnerUfa committed:
Итер3 проблема сторонних пакетов

SinnerUfa committed:
Итер3 проблема сторонних пакетов

SinnerUfa committed:
Итер3 проблема ошибок

SinnerUfa committed:
Итер3 проблема ошибок

SinnerUfa committed:
Итер3 проблема ошибок

SinnerUfa committed:
Итер3 проблема главной страницы

SinnerUfa committed:
Итер4 пробуем тесты

SinnerUfa committed:
Итер5 очистка комментариев, пробуем тесты

SinnerUfa committed:
Итер5 исправление имени окружения, пробуем тесты

SinnerUfa committed:
Спринт 1  - Исправление замечаний ревью 1
Замечания:
1. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1635381233]
2. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1635381396]

Исправления:
1. Удален блок кмментариев
2. Сделано логирование и возврат ошибки

SinnerUfa committed:
Спринт 1  - Исправление замечаний ревью 1
Замечания:
1. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1635384982]

Исправления:
1. Вставлен забытый возврат из функции

SinnerUfa committed:
Спринт 1  - Исправление замечаний ревью 1
Замечания:
1. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1635385943]
2. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1635377452]

Исправления:
1. Удален блок WaitGroup (данный блок от собственного шаблона сервиса)
2. Удален блок WaitGroup (данный блок от собственного шаблона сервиса)

SinnerUfa committed:
Спринт 1  - Исправление замечаний ревью 1
Замечания:
1. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1635374797]
2. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1635375062]
3. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1635375496]

Исправления:
1, 2, 3.
    - Методы логгера Crit и Error приведены к названиям методо стандартного логера (Fatal и Panic соотвественно) и исправлены все вызовы;
    - Вызовы метода логгера Eror изменены на Warning + возвраты из функций;
    - Ошибка сервера помещена в канал и передается на более высокий уровень;
    - Ошибки сервера внесены в общий список ошибок;

SinnerUfa committed:
Спринт 2 - Исправление замечаний ревью 2

Замечания:
1. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1642432834]
2. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1642433200]
3. [https://github.com/SinnerUfa/practicum-metrics/pull/5#discussion_r1642434440]

Исправления:
1, 2
    - Вызовы log.Warning(err) + os.Exit(1) заменены на log.Fatal(err)
3
    - Исключены комментарии

SinnerUfa committed:
Спринт 2 - Ввод теста для хандлеров (в составе сервера)

SinnerUfa committed:
Спринт 2 - Исправлена ошибка обработки строк выявленная при добавлении теста хандлеров

SinnerUfa committed:
Спринт 2 - Предыдущий коммит содержал ошибку обрезки строки

SinnerUfa committed:
Спринт 2 - расширение тестов сервера

SinnerUfa committed:
Спринт 2 - изменение префикса сообщений логера

SinnerUfa committed:
Спринт 2 - изменение логера на slog
- log и его использование через интерфейс заменено на использование slog по ссылке
- изменены сообщения об ошибках
- сторонние логеры теперь можно подключать через slog
- лог ведется в один файл (out.log) в формате JSON

SinnerUfa committed:
Спринт 2 - изменения по проверке vet

SinnerUfa committed:
Спринт 2 - временный коммит с подключенным zap логером, сделать завтра сутра до итер 8

SinnerUfa committed:
Спринт 2 -Добавление  логера в mid

SinnerUfa committed:
Спринт 2 - Добавление обработчиков json

SinnerUfa committed:
Спринт 2 - коррекция замечаний go vet

SinnerUfa committed:
Спринт 2 - коррекция замечаний go vet

SinnerUfa committed:
Спринт 2 модификация логирования в мид и исправлениие на нил неформируемых адресов в метриках

SinnerUfa committed:
Спринт 2 - Добалввение поддержки сжатия

SinnerUfa committed:
Спринт 2 -исправление теста

SinnerUfa committed:
Спрнт 2 - Декомпрессор

SinnerUfa committed:
test

SinnerUfa committed:
test

SinnerUfa committed:
test

SinnerUfa committed:
test

SinnerUfa committed:
test

SinnerUfa committed:
test

SinnerUfa committed:
test

SinnerUfa committed:
test

SinnerUfa committed:
Спринт 2 - Удаление коментариев. В модуле декомпрессора специально оставлено чтобы показать что делалось...

SinnerUfa committed:
Спринт 2 - вроде сделал ...

SinnerUfa committed:
test

SinnerUfa committed:
test

SinnerUfa committed:
Спринт 2 - изменено странное название, в тикер добавлен тик при закрытии (штатном), убран делитель частоты (костыль)

SinnerUfa committed:
Спринт 2 -проверка что проблема именно в частоте (возвращен делитель частоты). Тик в локагх при перзапуске необнаружен вероятнее всего неверно использован контекст

SinnerUfa committed:
Revert "Спринт 2 -проверка что проблема именно в частоте (возвращен делитель частоты). Тик в локагх при перзапуске необнаружен вероятнее всего неверно использован контекст"

This reverts commit 8d5bc0bacbafab9a3dae9c1ac24c514e434fb48a.

SinnerUfa committed:
Спринт 10 начало

SinnerUfa committed:
Спринт 10 - Убрал mutex надо подумать как их сделать правильно

SinnerUfa committed:
Спринт 3 - устранение замечаний предыдущего спринта

SinnerUfa committed:
Спринт 3 - убрал работу с декомпрессором и компрессором в формирователь хандлера, исключил декомпрессор из мид

SinnerUfa committed:
Спринт 3 - правка тестов

SinnerUfa committed:
Спринт 3 - правка вет

SinnerUfa committed:
Спринт 3 - правка вет, как выяснилось 9 итер все еще валится (ЭТО НУЖНО ПОПРАВИТЬ!)

SinnerUfa committed:
Cghbyn 3 - попытка найти ошибку

SinnerUfa committed:
Спринт 3 - попытка найти ошибку

SinnerUfa committed:
поиск ошибки

SinnerUfa committed:
Переделка сохранения в файл

SinnerUfa committed:
Переделка сохранения в файл

SinnerUfa committed:
Переделка сохранения в файл

SinnerUfa committed:
Переделка сохранения в файл

SinnerUfa committed:
Переделка сохранения в файл

SinnerUfa committed:
Переделка сохранения в файл

SinnerUfa committed:
Переделка сохранения в файл 2 из 10 проверок прохождения итер 9

SinnerUfa committed:
Переделка сохранения в файл 3 из 10 проверок прохождения итер 9

SinnerUfa committed:
Переделка сохранения в файл 4 из 10 итер 9 (4 провал)

SinnerUfa committed:
Переделка сохранения в файл 5 из 10 итер 9 (4 провал)

SinnerUfa committed:
1 Изменен логер, передача логера через параметры функции убрана
2 Убран пакет jmetric, представление метрики в json реализовано в пакете метрик
3 Темплейты убраны в отдельную папку
4 Поправлены уровни сообщений логера
5 Загрузка метрик перенесена в пакет загруки агента
TODO:
1 Убрать из пакета метрик типы gauge и counter (нужны только  storage) и кроме сеттера видимо сделать аддер для всех типов (тут надо подумать, где логика что делать с данными когда нет разделения при получении)
2 Разделить репозиторий и хранилище (repository и storage соответственно)
3 Для репозитория реализовать методы создания и закрытия
4 Для хранилищ (в памяти, файле, бд) реализовать соответствующие методы

SinnerUfa committed:
1 Поправлена ошибка Unmarshal

SinnerUfa committed:
1 Перемещены типы counter и gauge в хранилище (в памяти == файловое)

SinnerUfa committed:
1 Пакеты работы с переменными окружения и флагами перемещены в пакет конфигурации

SinnerUfa committed:
1 В метрики введены обшщие интерфейсы и все обращения делаются через них без привязки к репозиторию
2 Репозиторий разделен (временная версия)
3 Удален пустой хендлер, и ненужные readme
4

SinnerUfa committed:
1 Правка теста

SinnerUfa committed:
1 Добавлена функция закрытия репозитория, и его реализация в файловом хранилище
2 Удалены некоторые коды ошибок
3 Добавлено ожидание закрытия репозитория в серверной части

SinnerUfa committed:
1 Забыл раскомментировать

SinnerUfa committed:
1 Реализовано нормальное завершение сервисов

SinnerUfa committed:
1 Добавлен шаблон пакета для работы с БД
2 Реализован пинг
TODO: необходимо реализовать все интерфейсы репозитория с контекстом

SinnerUfa committed:
1 Расширен интерфейс репозитория функциями с контекстом (нужно для работы с БД)
2 Реализованы адаптеры к хранилищам в памяти и файле
TODO: Придется снова просить отсрочку ~4 дня (чтобы наверняка)

SinnerUfa committed:
1 Добавлен коммент для создания PR

SinnerUfa committed:
1 Реализовано хранение метрик в бд (моки не реализованы)
TODO: Необходимо вставить контексты в обработчики где они требуются

SinnerUfa committed:
1 Правка по Go Vet

SinnerUfa committed:
1 Поиск ошибки, скорее всего подготовка запросов не поддерживает @запись

SinnerUfa committed:
1 Переделка запросов в формат Postrage

SinnerUfa committed:
1 Забытая инициализация карты

SinnerUfa committed:
1 Убран проверочный лог

SinnerUfa committed:
1 Добавлен  хандлер загрузки батча

SinnerUfa committed:
1 Тест итера 12 без проверки клиента (но все же надо добавить)
2 Добавлено по-умолчанию отправка батчами (изменен конфиг), в слчучае если не будут проходить более ранние тесты то выключить конфиг

SinnerUfa committed:
1 Добавлены повторы отправки при ошибках
2 В части сервера такие повторы по-моему неуместны, клиент должен при ошибке решать надо или нет делать повтор

tmvrus commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,39 @@
+	mlog "github.com/SinnerUfa/practicum-metric/internal/mlog"
+)
+
+var cfg = agent.DefaultConfig
```
зачем тебе глобальная переменная? все равно же ты ее передаешь чем аргументы, так что смысла нет.

SinnerUfa replied:
Поправлено

tmvrus commented on cmd/server/main.go:
```diff
@@ -1,3 +1,39 @@
+	server "github.com/SinnerUfa/practicum-metric/internal/server"
+)
+
+var cfg server.Config = server.DefaultConfig
```
и тут тоже самое

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus commented on internal/agent/poster.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+func (m *MetricPost) Post() (err error) {
+	l, _ := m.rep.GetList()
```
а чего обработка ошибки игнорируется ?

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus commented on internal/agent/poster.go:
```diff
@@ -0,0 +1,70 @@
+	if m.noBatch {
+		endpoint := "http://" + m.adress + "/update/"
+		for i, v := range l {
+			req := client.R().SetHeader("Content-Type", "application/json").SetHeader("Content-Encoding", "gzip").SetBody(v)
+			p, err := req.Post(endpoint)
+			slog.Debug("", "req.Body", req.Body)
+			slog.Debug("", "content-encoding", p.Header().Get("Content-Encoding"), "response", p.String())
+			if err != nil {
+				slog.Warn("", "err", err, "i", i, "value", v)
+			}
+		}
+	} else {
+		endpoint := "http://" + m.adress + "/updates/"
+		req := client.R().SetHeader("Content-Type", "application/json").SetHeader("Content-Encoding", "gzip").SetBody(l)
+		p, err := req.Post(endpoint)
+		slog.Debug("", "req.Body", req.Body)
+		slog.Debug("", "content-encoding", p.Header().Get("Content-Encoding"), "response", p.String())
+		if err != nil {
+			slog.Warn("", "err", err, "l", l)
+		}
```
весьма странный сообщения логирования

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus commented on internal/agent/poster.go:
```diff
@@ -0,0 +1,70 @@
+func (m *MetricPost) Tick() {
+	var delay int
+
+	for counter := 0; counter < 3; counter++ {
+		err := m.Post()
+		switch counter {
+		case 0:
+			delay = 1
+		case 1:
+			delay = 3
+		case 2:
+			delay = 5
+		}
+		if err != nil {
+			time.Sleep(time.Duration(delay) * time.Second)
+		} else {
+			break
+		}
+	}
```
выглядит как-то очень сложно, почему просто не переписать цикл
```
for _, delay := range []time.Duration{time.Second, 3*time.Second, 5*time.Second} {
    
}
```

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus commented on internal/repository/database/database.go:
```diff
@@ -0,0 +1,264 @@
+const MaxConns int = 10
+
+const (
+	CreateTableCounters string = `CREATE TABLE IF NOT EXISTS counters (
+                                        cnt_name character varying NOT NULL,
+                                        cnt_value bigint NOT NULL,
+                                        CONSTRAINT cnt_name_prim PRIMARY KEY (cnt_name)
+                                    );`
+
+	InsertIntoCounters string = ` INSERT INTO counters ( cnt_name, cnt_value ) VALUES ( $1, $2 )
+                                    ON CONFLICT ON CONSTRAINT cnt_name_prim DO
+                                    UPDATE SET cnt_value = counters.cnt_value + EXCLUDED.cnt_value;`
+
+	SelectAllCounters  string = `SELECT counters.cnt_name, counters.cnt_value FROM counters;`
+	SelectNameCounters string = `SELECT counters.cnt_value FROM counters WHERE counters.cnt_name = $1 LIMIT 1;`
+
+	CreateTableGauges string = `CREATE TABLE IF NOT EXISTS gauges (
+                                    gau_name character varying NOT NULL,
+                                    gau_value double precision NOT NULL,
+                                    CONSTRAINT gau_name_prim PRIMARY KEY (gau_name)
+                                );`
+
+	InsertIntoGauges string = `INSERT INTO gauges (gau_name, gau_value) VALUES ($1, $2)
+                                ON CONFLICT ON CONSTRAINT gau_name_prim DO
+                                UPDATE SET gau_value = EXCLUDED.gau_value;`
+
+	SelectAllGauges  string = `SELECT gauges.gau_name, gauges.gau_value FROM gauges;`
+	SelectNameGauges string = `SELECT gauges.gau_value FROM gauges WHERE gauges.gau_name = $1 LIMIT 1;`
```
`string` можно не писать, компилятор достаточно умный что бы вывести тип в данном случае

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus commented on internal/repository/database/database.go:
```diff
@@ -0,0 +1,264 @@
+func New(ctx context.Context, dsn string) (*Database, error) {
+	db, err := sql.Open("pgx", dsn)
+	if err != nil {
+		return nil, errors.Join(err, db.Close())
```
`db.Close()` встречается аж 4 раза в этом коде, можно перписать чуть проще,
 ```
 New(ctx context.Context, dsn string) (DB *Database, err error) {
     defer func(){
       if err != nil {
           err = errors.Join(err, db.Close())
       }
     }
 }
``` 
это если обязательно хочется ошибку закрытия воззвращать, но обычно так не делают, а просто ее логируют в том же defer

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus commented on internal/repository/database/database.go:
```diff
@@ -0,0 +1,264 @@
+	return DB, nil
+}
+
+func (d *Database) SetContext(ctx context.Context, m metrics.Metric) error {
```
это какая-то сложная вещь, почему ф-я `SetContext` вместо установки контекста(что само по себе довольно странно, ну ок)
- создает транзакцию
- выполняет в ней 1 запрос (зачем для одного запроса вообще транзакция)

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus commented on internal/repository/database/database.go:
```diff
@@ -0,0 +1,264 @@
+	}
+}
+
+func (d *Database) GetContext(ctx context.Context, m *metrics.Metric) error {
```
вот тут тоже большой вопрос к именованию

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus commented on internal/repository/database/database.go:
```diff
@@ -0,0 +1,264 @@
+}
+
+func (d *Database) GetContext(ctx context.Context, m *metrics.Metric) error {
+	if m.Name == "ping" && m.Type == "" && m.Value.IsString() && m.Value.String() == "ping" {
```
зачем смешивать служебный Ping и продуктовые метрики?

SinnerUfa replied:
Тут  у меня большой вопрос как правильно репозиторий общий с общим интерфейсом для всех сторедж. Пинг как то вообще вываливается по смыслу. Куда его добавлять?  прокидывать db в хандлер?

tmvrus replied:
сделай отдельный сторадж который никак не связан с продуктовым кодом

tmvrus replied:
Физичический конект к базе можно передавать в любое количество стораджей, на работу пула конектов это не повлияет.

tmvrus commented on internal/repository/database/database.go:
```diff
@@ -0,0 +1,264 @@
+		return errors.Join(err, tx.Rollback())
+	}
+
+	return tx.Commit()
```
а кто сделает rollback?

SinnerUfa replied:
Замечание некорректно откат в формировании ошибки

tmvrus replied:
> Замечание некорректно откат в формировании ошибки

Прошу прощения, не заметил.

tmvrus commented on internal/server/hundlers/get_list.go:
```diff
@@ -0,0 +1,74 @@
+			}
+			t.Funcs(funcMap)
+			t, err := t.Parse(tpi)
+			// t, err := t.ParseFiles("body_list.html", "index.html") // NOT WORK WITH FUNCS???
```
вот не надо закоменченного кода в рабочем проекте, замусоривает и усложняет чтение

SinnerUfa replied:
Поправлено (отмечаю т.к. пуши делаются руками)

tmvrus reviewed (changes requested):
Идея контекста, подразумевает что он будет пробрасываться от входа с приложение(хендлер), до выхода(база). Создавать `context.Background()` на промежуточном слое, разрывает связь между входом и выходом, а это не имеет никакого смысла.

SinnerUfa committed:
Add files via upload

SinnerUfa committed:
Исправление обработки ошибки (хотя она всегда 0)

SinnerUfa committed:
Возврат кода без ошибки,  перестал проходить тест

Возврат кода без ошибки,  перестал проходить тест

SinnerUfa committed:
Возврат обработки ошибки

SinnerUfa committed:
Исправления по замечаниям

Исправление логирования, исправление повторов отправки
Снова внезапно прошел тест

SinnerUfa committed:
Исправление замечаний

Убрано опдеределние типа в const

SinnerUfa committed:
Исправление замечаний

Скорректирована функция открытия бд

SinnerUfa committed:
Исправление замечаний

Поправлено формирование err (было создание нового что при defer не будет работать)

SinnerUfa committed:
Исправление замечаний

Убран коммент

SinnerUfa committed:
Исправления замечаний

Поправлено наименование интерфесов с контекстом

SinnerUfa committed:
Ошибка

Исправление ошибки

SinnerUfa committed:
Странное поведение функции повтора

Странное поведение функции повтора

SinnerUfa committed:
Добавлена проверка на nil в defer

SinnerUfa committed:
Все интерфейсы через контекст

Все интерфейсы через контекст

SinnerUfa committed:
Забытая правка теста

Забытая правка теста

SinnerUfa commented:
Остался только вопрос по ping, и есть вероятность повтора ошибки теста итера 13, скорее всего правками она должна уйти но я не уверен

SinnerUfa commented:
Остался только вопрос по ping, и есть вероятность повтора ошибки теста итера 13, скорее всего правками она должна уйти но я не уверен

SinnerUfa commented:
По пингу есть вариант сделать в интерфейсе репозитория Status() bool или что-то подобное, и чрезе него прокинуть пинг, исхожу из того что никакой хендлер не должен знать с каким репозиторием он работает.

SinnerUfa committed:
Вывод пинга из прода

Криво наверное но должно работать

SinnerUfa committed:
Забыл про тесты

Поправлен файл теста

SinnerUfa committed:
Забыл и фалй сервера...

Пора спать похоже

tmvrus commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,43 @@
+	"os"
+	"os/signal"
+
+	agent "github.com/SinnerUfa/practicum-metric/internal/agent"
+	config "github.com/SinnerUfa/practicum-metric/internal/config"
+	mlog "github.com/SinnerUfa/practicum-metric/internal/mlog"
```
довольно странно давать названием модулей алиасы которые такие же как имена

SinnerUfa replied:
Если честно привычка, ничего в этом полезного нет, но вредного тоже, могу поправить без проблем)

SinnerUfa replied:
А  пинг все равно как костыль, не нравится мне, но лучше пока не придумал) За статус, вроде бы, вообще другое апи должно отвечать)

tmvrus replied:
> Если честно привычка, ничего в этом полезного нет, но вредного тоже, могу поправить без проблем)

есть такой полезный методологический принцип "бритва оккама", можно его сформулировать как "не делайте ничего что не нужно" в данном случае, алиасы ничего не добавляют) а значит противоречат этому принципу)

tmvrus commented on internal/agent/poster.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+func (m *MetricPost) Tick() {
+	delays := []int{1, 3, 5}
```
можно было сразу сделать кстати:
```
delays := []int{time.Second, 3*time.Second, 5*time.Second}
```

SinnerUfa replied:
Да так и было, верну. На этом коммите был завал теста, но причина в другом конечно была. Спасибо за проверку. Надеюсь послезавтра будет до 15го итера)

tmvrus reviewed (approved):
Ping получился в вполне себе.

SinnerUfa committed:
Обновление версии go в тестах 

Обновление версии go в тестах до 1.23 (правильность действий сомнительна)
